File: iruri/lib/model/article.dart
import 'dart:convert';

/**
 *  image는 데이터베이스에서 가져올 때 Buffer가 저장되는 곳 입니다.
 *  imagePath는 post할 때 데이터베이스에 저장되는 경로입니다.
 *  DateTime을 모두 String으로 변경하였습니다.
 *  Period class가 추가되었습니다.
 * 
 *  @seunghwanly 
 */
class Article {
  final String id; // needs to be converted to <ObjectID>
  final List<String> members;
  final List<String> contracts;
  final Detail detail;
  final String imagePath;
  final String image;

  Article(
      {this.id,
      this.contracts,
      this.members,
      this.detail,
      this.imagePath,
      this.image});

  factory Article.fromJson(Map<String, dynamic> json) {
    return Article(
        id: json['_id'],
        image: json['image'],
        contracts:
            json['contracts'] != null ? List.from(json['contracts']) : [],
        members: json['members'] != null ? List.from(json['members']) : [],
        imagePath: json['imagePath'],
        detail: Detail.fromJson(json['detail']));
  }

  // for posting new article
  Map<String, dynamic> toJson() => {
        'imagePath': imagePath,
        'members': members,
        'contracts': contracts,
        'detail': detail.toJson()
      };
}

// Detail Class
class Detail {
  final String status;
  // final DateTime reportedDate;
  // final DateTime dueDate;
  final String reportedDate;
  final String dueDate;
  final Period period;
  final Condition condition;
  final Content content;
  final String writer;
  final String location;
  final Applicant applicant;

  Detail(
      {this.applicant,
      this.condition,
      this.content,
      this.dueDate,
      this.location,
      this.period,
      this.reportedDate,
      this.status,
      this.writer});

  factory Detail.fromJson(Map<String, dynamic> json) => Detail(
      period: Period.fromJson(json['period']),
      applicant: Applicant.fromJson(json['applicant']),
      reportedDate: json['reportedDate'],
      dueDate: json['dueDate'],
      status: json['status'],
      condition: Condition.fromJson(json['condition']),
      content: Content.fromJson(json['content']),
      writer: json['writer'],
      location: json['location']);

  Map<String, dynamic> toJson() => {
        'period': period.toJson(),
        'applicants': applicant.toJson(),
        // 'reportedDate': DateFormat("yyyy-MM-ddTHH:mm:ss").format(reportedDate),
        // 'dueDate': DateFormat("yyyy-MM-ddTHH:mm:ss").format(dueDate),
        "reportedDate": reportedDate,
        "dueDate": dueDate,
        'status': status,
        'condition': condition.toJson(),
        'content': content.toJson(),
        'writer': writer,
        'location': location
      };
}

// Condition Class
class Condition {
  final String projectType;
  final String contractType;
  final String wage;

  Condition({this.contractType, this.projectType, this.wage});

  factory Condition.fromJson(Map<String, dynamic> json) => Condition(
      contractType: json['contractType'],
      projectType: json['projectType'],
      wage: json['wage']);

  Map<String, String> toJson() =>
      {'contractType': contractType, 'projectType': projectType, 'wage': wage};
}

// Content Class
class Content {
  final String title;
  final String desc;
  final List<String> tags;
  final List<String> genres;
  final String prefer;

  Content({this.desc, this.title, this.prefer, this.tags, this.genres});

  factory Content.fromJson(Map<String, dynamic> json) => Content(
        tags: List.from(json['tags']),
        genres: List.from(json['genres']),
        desc: json['desc'],
        prefer: json['prefer'],
        title: json['title'],
      );

  Map<String, dynamic> toJson() => {
        'tags': tags,
        'genres': genres,
        'desc': desc,
        'prefer': prefer,
        'title': title
      };
}

class Period {
  // final DateTime from;
  // final DateTime to;
  final String from;
  final String to;

  Period({this.from, this.to});

  factory Period.fromJson(Map<String, dynamic> json) => Period(
      // from: DateTime.parse(json['from']), to: DateTime.parse(json['to'])
      from: json['from'],
      to: json['to']);

  Map<String, String> toJson() => {
        // 'from': DateFormat("yyyy-MM-ddTHH:mm:ss").format(from),
        // 'to': DateFormat("yyyy-MM-ddTHH:mm:ss").format(to)
        "from": from,
        "to": to
      };
}

class Applicant {
  final List<String> writeMains; // 메인글
  final List<String> writeContis; // 글콘티
  final List<String> drawMains; // 메인그림
  final List<String> drawContis; // 그림콘티
  final List<String> drawDessins; // 그림뎃셍
  final List<String> drawLines; // 선화
  final List<String> drawChars; // 캐릭터
  final List<String> drawColors; // 채색
  final List<String> drawAfters; // 후보정

  Applicant(
      {this.drawAfters,
      this.drawChars,
      this.drawColors,
      this.drawContis,
      this.drawDessins,
      this.drawLines,
      this.drawMains,
      this.writeContis,
      this.writeMains});

  factory Applicant.fromJson(Map<String, dynamic> json) => Applicant(
        writeMains: List<String>.from(json['writeMains']),
        writeContis: List<String>.from(json['writeContis']),
        drawMains: List<String>.from(json['drawMains']),
        drawContis: List<String>.from(json['drawContis']),
        drawDessins: List<String>.from(json['drawDessins']),
        drawChars: List<String>.from(json['drawChars']),
        drawColors: List<String>.from(json['drawColors']),
        drawAfters: List<String>.from(json['drawAfters']),
        drawLines: List<String>.from(json['drawLines']),
      );

  Map<String, dynamic> toJson() => {
        "writeMains": jsonEncode(writeMains),
        "writeContis": jsonEncode(writeContis),
        "drawMains": jsonEncode(drawMains),
        "drawContis": jsonEncode(drawContis),
        "drawDessins": jsonEncode(drawDessins),
        "drawChars": jsonEncode(drawChars),
        "drawColors": jsonEncode(drawColors),
        "drawAfters": jsonEncode(drawAfters),
        "drawLines": jsonEncode(drawLines)
      };
}

File: server/models/article.js
const mongoose = require('mongoose');

// define schema
// sub document
const conditionSchema = new mongoose.Schema({
    projectType: String,
    contractType: String,
    wage: String
});
// sub document
const contentSchema = new mongoose.Schema({
    title: String,
    desc: String,
    tags: [String],
    genres: [String],
    prefer: String
});
const periodSchema = new mongoose.Schema({
    from: { type: Date, default: Date.now },
    to: { type: Date, default: Date.now }
});
/**
 * status for applicant in article
    * @enum {number}
    * 승인대기 : 0
    * 승인거절 : -1
    * 승인완료 : 1
    * 계약서대기 : 2
    * 계약서작성완료 : 3
 */
const applicantSchema = new mongoose.Schema({
    uid: { type: String, required: true },
    status: { type: Number, required: true }
});
// middle document
const detailSchema = new mongoose.Schema({
    status: String,
    reportedDate: {
        type: Date,
        default: Date.now
    },
    dueDate: {
        type: Date,
        default: Date.now
    },
    period: {
        _id: false,
        type: periodSchema
    },
    condition: {
        // don't create objectID in sub document
        _id: false,
        type: conditionSchema
    },
    content: {
        // don't create objectID in sub document
        _id: false,
        type: contentSchema
    },
    writer: String,
    location: String,
    /* 
        지원자를 포지션 별로 담아 낼 수 있어야 하기 때문에 구조 변화를 주어야함
        applicants : {
            write_mains : []
            write_contis : []
            draw_mains : []
            draw_contis : []
            draw_dessins : []
            draw_lines : []
            draw_chars : []
            draw_colors : []
            draw_afters : []
        }
    */
    applicant: {
        writeMains: [{ type: applicantSchema, _id: false }],
        writeContis: [{ type: applicantSchema, _id: false }],
        drawMains: [{ type: applicantSchema, _id: false }],
        drawContis: [{ type: applicantSchema, _id: false }],
        drawDessins: [{ type: applicantSchema, _id: false }],
        drawLines: [{ type: applicantSchema, _id: false }],
        drawChars: [{ type: applicantSchema, _id: false }],
        drawColors: [{ type: applicantSchema, _id: false }],
        drawAfters: [{ type: applicantSchema, _id: false }],
    }
});
// parent document
const articleSchema = new mongoose.Schema({
    // 💡 MongoDB does not allow overwriting the default _id
    imagePath: String,
    image: String,
    members: [{ type: String }],
    contracts: [{ type: String }],
    detail: {
        type: detailSchema,
        // don't create objectID in sub document
        _id: false
    }
});

// export model
module.exports = mongoose.model('Article', articleSchema);